docs/design/provisioning.md

```diff
@@ -0,0 +1,328 @@
+// within. The Recycler controller running in master will use the specified plugin and optional arguments.
+// Recyclers can be plugins that delete cloud volumes through APIs or pods that scrub a volume for reuse.
+// Each Recycler plugin requires its own documentation for usage and plugin arguments.
+type RecyclerConfig struct {
+	// PersistentVolumeSelector is a selector which must be true for the recycler to match the persistent volume
+	PersistentVolumeSelector Selector
+	// RecyclerPluginName is the name of the plugin that can recycle the matched persistent volume.
+	RecyclerPluginName string
+	// Args is a map of key/value pairs passed to the driver (but not plugin) as arguments
+	Args *map[string]string
+	// SecurityContextName is the name of the security context to use when performing recycling operations
+	SecurityContextName string
+}
+
+// ProvisionerConfig is a label set for persistent volume claims, a plugin w/ optional arguments,
+// and a security context to work within.  The provisioning controller running in master will use the specified
+// plugin to provision volumes for claims that match the selector.
+// Each Provisioner plugin requires its own documentation for usage and plugin arguments.
+type ProvisionerConfig struct {
+	// PersistentVolumeClaimSelector is a selector which must be true for the provisioner to match the claim
+	PersistentVolumeClaimSelector Selector
+	// ProvisionerPluginName is the name of the plugin that will provision the claim matched by the selector
+	ProvisionerPluginName string
+	// Args is a map of key/value pairs passed to the driver (but not plugin) as arguments
+	Args *map[string]string
+	// SecurityContextName is the name of the security context to use when performing provisioning operations
+	SecurityContextName string
+}
+
+```
+
+### Provisioning Plugins
+
+New plugins:
+
+`PersistentVolumeProvisioner` is a plugin that can create real storage with a storage provider (such as EBS or GCE).
+
+`PersistentVolumeRecycler` is a plugin that reclaims discarded resources.
+
+> The Deleter plugin and interfaces to be removed. Deletion is one of many possible Recycling outcomes.
+> The Creater plugin and interfaces to be refactored to Provisioner
+
+
+Provisioning and Recycling both follow a plugin model where the plugin accepts  generic arguments.  A future enhancement can scan for drivers by name and consume the same arguments (following a similar pattern in  [Networking](../../pkg/kubelet/network/exec/exec.go)).
+
+The plugin model is similar to volume plugins, where each has a unique name and implements various interfaces.  Arguments as passed as string key/value pairs to a plugin in `Init(host, args)`, which will be called exactly once before any plugin operation is invoked.  Many provisioners can be configured using the same plugin with different arguments, such as "volumeType" for a Cloud API.
```

Comment:
I'm not yet sold on the model proposed here.  Volumes suffer from a serious problem today - they aren't extensible except via checking code into Kube.  A provisioner and a binder are controllers - the details of how they bind are really their own domain.  An admin configures their applicability - that problem is solely the domain of administrators.

I'm going to describe an alternate model that I think is superior:

1. We treat binders and provisioners as sets of controllers
2. The implementation of those controllers is shared, and may be in the Kube source tree or another
3. It must be easy for someone to glue together their own provisioner or binder with or without writing Go code
4. When someone wants to write a new provisioner, they simply watch for PVCs based on some criteria
5. If they are going to provision for the PVC, they annotate the PVC first (to take ownership) and then create a PV, then satisfy the PVC
6. There should be a generic provisioner controller that can shell out to bash.  The arguments to bash are details about the PVC.  The arguments to the controller are a set of PVC labels to watch.  The bash command is responsible for provisioning a PV and then returning 0 and the name of the new PV, or 1 and the exit code.
7. There can be in-tree provisioners that replace the call out to bash with "call out to Go code compiled with me" via the interfaces described here.

The model above defines how anyone (not just a Kube programmer) can easily extend the logic.  An admin can separate those controllers by their label matchers.

Kubernetes is intended to be a compositional system - whenever we add new abstractions, we have to clear a pretty high bar that the abstraction is easily composible.  I think the proposal here does not go far enough to focus on that composability and extensibility.